cpp/src/table/table_device_view.cu

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <utilities/error_utils.hpp>
+
+#include <rmm/rmm.h>
+
+#include <algorithm>
+#include <numeric>
+
+namespace cudf {
+namespace detail {
+
+template <typename ColumnDeviceView, typename HostTableView>
+void table_device_view_base<ColumnDeviceView, HostTableView>::destroy() {
+  RMM_TRY(RMM_FREE(_columns, _stream));
+  delete this;
+}
+
+template <typename ColumnDeviceView, typename HostTableView>
+table_device_view_base<ColumnDeviceView, HostTableView>::table_device_view_base(
+    HostTableView source_view, cudaStream_t stream)
+    : _num_rows{source_view.num_rows()},
+      _num_columns{source_view.num_columns()},
+      _stream{stream} {
+  if (source_view.num_columns() > 0) {
+    size_type total_descendants =
+        std::accumulate(source_view.begin(), source_view.end(), 0,
+                        [](size_type init, column_view col) {
+                          return init + count_descendants(col);
+                        });
+    CUDF_EXPECTS(0 == total_descendants,
+                 "Columns with descendants are not yet supported.");
+
+    auto views_size_bytes =
+        source_view.num_columns() * sizeof(*source_view.begin());
+    RMM_TRY(RMM_ALLOC(&_columns, views_size_bytes, stream));
+    CUDA_TRY(cudaMemcpyAsync(_columns, &(*source_view.begin()),
+                             views_size_bytes, cudaMemcpyDefault, stream));
+  }
+}
```

Comment:
This code seems a little risky. `table_view` has `column_view's` and `table_device_view` has `column_device_view's`.
The `source_view` parameter here is going to be either `table_view` or `mutable_table_view` per the `HostTableView` template type. This constructor is to create an array of `column_device_view's` in the `_columns` member from the `column_view's` inside of `source_view`. Looks like you are doing this by memcpy'ing the `column_view's` into the `_columns` member (line 54). You are assuming that `column_view === column_device_view`. These are definitely not the same.
I think you need to do something like this:
```
    auto views_size_bytes = _num_columns_ * sizeof(ColumnDeviceView);
    std::vector<int8_t> h_buffer(views_size_bytes); // allocate host memory
    RMM_TRY(RMM_ALLOC(&_columns, views_size_bytes, stream)); // allocate device memory
    ColumnDeviceView* h_column = reinterpret_cast<ColumnDeviceView*>(h_buffer.data());
    for( size_type idx=0; idx < _num_columns; ++idx )
    {
      // inplace-new the column_device_view object from the column_view object
      new(h_column) ColumnDeviceView(source_view.column(idx));
      h_column++;
    }
    CUDA_TRY(cudaMemcpyAsync(_columns, h_buffer.data(),
                             views_size_bytes, cudaMemcpyDefault, stream));
```
This will create each of the `ColumnDeviceView` objects in host memory in an array that can be copied over to device memory.

Reply:
Great catch! You are absolutely correct.